polkadot/runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -1029,3 +1033,125 @@ fn limit_and_sanitize_disputes<
+
+	// Process all backed candidates.
+	let backed_len_before = backed_candidates.len();
+	backed_candidates.retain_mut(|bc| {
+		// Get `core_idx` assigned to the `para_id` of the candidate (step 2)
+		let core_idx = match para_id_to_core_idx.get(&bc.descriptor().para_id) {
+			Some(core_idx) => *core_idx,
+			None => {
+				log::debug!(target: LOG_TARGET, "Can't get core idx got a backed candidate for para id {:?}. Dropping the candidate.", bc.descriptor().para_id);
+				return false
+			}
+		};
+
+		// Get relay parent block number of the candidate. We need this to get the group index
+		// assigned to this core at this block number
+		let relay_parent_block_number = match <shared::Pallet<T>>::allowed_relay_parents()
+			.acquire_info(bc.descriptor().relay_parent, None) {
+				Some((_, block_num)) => block_num,
+				None => {
+					log::debug!(target: LOG_TARGET, "Relay parent {:?} for candidate is not in the allowed relay parents. Dropping the candidate.", bc.descriptor().relay_parent);
+					return false
+				}
+			};
+
+
+		// Get the group index for the core
+		let group_idx = match <scheduler::Pallet<T>>::group_assigned_to_core(
+			core_idx,
+			relay_parent_block_number + One::one(),
+		) {
+			Some(group_idx) => group_idx,
+			None => {
+				log::debug!(target: LOG_TARGET, "Can't fetch group index for core idx {:?}. Dropping the candidate.", core_idx);
+				return false
+			},
+		};
+
+		// And finally get the validator group for this group index
+		let validator_group = match <scheduler::Pallet<T>>::group_validators(group_idx) {
+			Some(validator_group) => validator_group,
+			None => {
+				log::debug!(target: LOG_TARGET, "Can't get the validators from group {:?}. Dropping the candidate.", group_idx);
+				return false
+			}
+		};
+
+		// `validator_indices` in `BackedCandidate` are indecies within the validator group. Convert
+		// them to `ValidatorId`.
+		let voted_validator_ids = bc
+			.validator_indices
+			.iter()
+			.enumerate()
+			.filter_map(|(idx, bitval)| match *bitval {
+				true => validator_group.get(idx), // drop indecies not found in the validator group. This will lead to filtering the vote
+				false => None,
+			})
+			.collect::<Vec<_>>();
+
+		let validity_votes = std::mem::take(&mut bc.validity_votes);
+		let (validity_votes, dropped) : (Vec<(usize, ValidityAttestation)>, Vec<(usize, ValidityAttestation)>) =  validity_votes.into_iter().enumerate().partition(|(idx, _)| {
+			let voted_validator_index = match voted_validator_ids.get(*idx) {
+				Some(validator_index) => validator_index,
+				None => {
+					log::debug!(target: LOG_TARGET, "Can't find the voted validator index {:?} in the validator group. Dropping the vote.", group_idx);
+					return false
+				}
+			};
+
+			!disabled_validators.contains(voted_validator_index)
+		});
+
+		if !dropped.is_empty() {
+			filtered = true;
+		}
+
+		if validity_votes.is_empty() {
+			//everything is filtered out - remove the whole candidate
+			return false
```

Comment:
we will not only drop candidates with zero votes, but also with insufficient votes https://github.com/paritytech/polkadot-sdk/blob/21d36b7b4229c4d5225944f197918cde23fda4ea/polkadot/runtime/parachains/src/inclusion/mod.rs#L725-L737 which is why we should not duplicate the check logic

Reply:
I couldn't figure out a more elegant way to handle this. I fixed the code to check for `effective_minimum_backing_votes` instead of 0 votes.

Up to now the approach was "if client gives us bad data we drop it", which was fine. But with runtime filtering in place we need to be extra careful not to break any invariant for the data.

Do you think there is a benefit from an end-to-end test of `process_inherent_data`?